File: java/src/com/mazes/Distances.java
package com.mazes;

import java.util.HashMap;
import java.util.Set;

/**
 * Distances is a helper class that is used to hold distance information between a root cell and other cells
 * in a Grid.
 * When Distances is instantiated, a root cell must be specified to act as the origin point. Then use the put()
 * method to add Cells and their distances from the root cell.
 * These distances are used by path finding algorithms (like Dijkstra's)
 */
public class Distances {

    // the root cell, or origin, that distances will be computed from
    private final Cell root;

    // holds distances values for each Cell in a Grid
    private final HashMap<Cell, Integer> cells;

    public Distances(Cell root) {
        this.root = root;
        this.cells = new HashMap<>();
        this.cells.put(root, 0);
    }

    /**
     *
     * @return the root cell of this Distances object
     */
    public Cell getRoot() {
        return this.root;
    }

    /**
     * get the distance value for the specified cell from this Distance class' internal Map
     *
     * @param cell the Cell to get the distance for
     * @return an integer that indicates how far this cell is from a root cell. If the cell in not in this Distances
     * class then null is returned
     */
    public int get(Cell cell) {
        return this.cells.get(cell);
    }

    /**
     * @param cell
     * @return true if this Distance object contains an entry for the given cell, else false
     */
    public boolean contains(Cell cell) {
        return this.cells.containsKey(cell);
    }

    /**
     * puts the provided cell and distance information into this Distances class internal Map
     *
     * @param cell     the Cell to put
     * @param distance the distance the cell is from the root Cell
     */
    public void put(Cell cell, int distance) {
        this.cells.put(cell, distance);
    }

    /**
     * @return a set view of the Cells contained in this Distances internal map
     */
    public Set<Cell> cells() {
        return this.cells.keySet();
    }



}

File: nodejs/src/Distances.js
/**
 * Distances is a helper class that holds distance information between a "root" cell and other "linked" cells.
 * It manages an internal Map that maps cell objects to number values, where the number is that cells
 * distance from the root cell.
 */
class Distances {

    /**
     * constructs a new Distances object with rootCell as the "root"
     * @param rootCell
     */
    constructor(rootCell) {
        this._root = rootCell;
        this._cells = new Map();
        this._cells.set(this._root, 0);
        this._rootCell = rootCell;
    }

    /**
     *
     * @returns {Cell} the root cell of this Distances object
     */
    get root() {
        return this._root;
    }

    /**
     * gets the distance for specified cell
     * @param cell
     * @returns {Number | undefined} the distance value for the specified cell. If the cell is not contained in this
     * distances object, undefined is returned
     */
    distanceOf(cell) {
        return this._cells.get(cell);
    }

    /**
     * checks if this Distances object has an entry for cell
     * @param cell {Cell} the cell to search for
     * @returns {boolean} true if the specified cell has an entry in this distances object, else false
     */
    contains(cell) {
        return this._cells.has(cell);
    }

    /**
     * add the specified cell and distance to this distance object
     * @param cell
     * @param distance
     */
    put(cell, distance) {
        this._cells.set(cell, distance);
    }

    /**
     *
     * @returns {IterableIterator<any>} an iterator over the cells contained in this Distances object
     */
    cells() {
        return this._cells.keys();
    }
}

module.exports = Distances;